api/webrtc/webrtc.go

```diff
@@ -0,0 +1,388 @@
+	})
+
+	offer, err := c.pc.CreateOffer(&webrtc.OfferOptions{ICERestart: false})
+	if err != nil {
+		return nil
+	}
+	if err = c.pc.SetLocalDescription(offer); err != nil {
+		return nil
+	}
+
+	c.dc = dc
+	return nil
+}
+
+func (c *Peer) Answer(offerSdp string) error {
+	offer := webrtc.SessionDescription{}
+	err := Decode(offerSdp, &offer)
+	if err != nil {
+		return err
+	}
+
+	sdp, err := offer.Unmarshal()
+	if err != nil {
+		return err
+	}
+	limiter := ratelimiter.GetLimiter("webrtc:"+sdp.Origin.UnicastAddress, config.Parameters.WsIPRateLimit, int(config.Parameters.WsIPRateBurst))
+	if !limiter.Allow() {
+		return fmt.Errorf("webrtc connection limit of %s reached", sdp.Origin.UnicastAddress)
+	}
+
+	if err := c.pc.SetRemoteDescription(offer); err != nil {
+		return err
+	}
+
+	answer, err := c.pc.CreateAnswer(nil)
+	if err != nil {
+		return err
+	}
+
+	if err := c.pc.SetLocalDescription(answer); err != nil {
+		return err
+	}
+
+	c.pc.OnICECandidate(func(candidate *webrtc.ICECandidate) {
+		if candidate == nil {
+			localDesc := c.pc.LocalDescription()
+			encodedDescr, err := Encode(localDesc)
+			if err != nil {
+				log.Error("WebRTC OnICECandidate error: ", err)
+				return
+			}
+			c.answer = encodedDescr
+			c.OnSdp <- encodedDescr
+		}
+	})
+
+	c.pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		dc.OnOpen(func() {
+			c.dc = dc
+
+			c.mutex.Lock()
+			c.isConnected = true
+			c.mutex.Unlock()
+
+			if NewConnection != nil {
+				go NewConnection(c, nil)
+			}
+		})
+
+		dc.OnMessage(func(msg webrtc.DataChannelMessage) {
+			var dcmsg DataChannelMessage
+			if msg.IsString {
+				dcmsg.messageType = TextMessage
+				if string(msg.Data) == PingData {
+					dc.SendText(PongData)
+					return
+				} else if string(msg.Data) == PongData {
+					if c.pongHandler != nil {
+						c.pongHandler(PongData)
+					} else {
+						log.Info("Pong handler not set")
+					}
+					return
+				}
+			} else {
+				dcmsg.messageType = BinaryMessage
+			}
+			dcmsg.data = msg.Data
+			c.OnMessage <- dcmsg
+		})
+
+		dc.OnClose(func() {
+			c.mutex.Lock()
+			c.isConnected = false
+			c.mutex.Unlock()
+		})
+
+		dc.OnError(func(err error) {
+			log.Errorf("Data Channel %v error: %s\n", dc.Label(), err.Error())
+		})
+	})
+
+	return nil
+}
+
+func (c *Peer) IsConnected() bool {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.isConnected
+}
+
+func (c *Peer) SetRemoteSdp(sdp string) error {
+	answer := webrtc.SessionDescription{}
+	err := Decode(sdp, &answer)
+	if err != nil {
+		return err
+	}
+	return c.pc.SetRemoteDescription(answer)
+}
+
+func (c *Peer) WriteMessage(messageType int, data []byte) (err error) {
+	if c.dc == nil {
+		return fmt.Errorf("DataChannel not available")
+	}
+	if data == nil {
+		if messageType == PingMessage {
+			return c.dc.SendText(PingData)
+		} else if messageType == PongMessage {
+			return c.dc.SendText(PongData)
+		}
+		return nil
+	}
+
+	writeResult := make(chan error)
+	go func() {
```

Comment:
write should only be called if write deadline has not passed yet

Reply:
Yes, let me modify it.